waku/node/v2/message_store.nim

```diff
@@ -0,0 +1,216 @@
+      discard sqlite3_finalize(journalModePragma)
+      return err($sqlite3_errstr(x))
+
+    if (let x = sqlite3_column_text(journalModePragma, 0);
+        x != "memory" and x != "wal"):
+      discard sqlite3_finalize(journalModePragma)
+      return err("Invalid pragma result: " & $x)
+
+  # TODO: check current version and implement schema versioning
+  checkExec "PRAGMA user_version = 1;"
+
+  let journalModePragma = prepare("PRAGMA journal_mode = WAL;"): discard
+  checkWalPragmaResult(journalModePragma)
+  checkExec(journalModePragma)
+
+  ## Table is the SQL query for creating the messages Table.
+  ## It contains:
+  ##  - 4-Byte ContentTopic stored as an Integer
+  ##  - Payload stored as a blob
+  checkExec """
+    CREATE TABLE IF NOT EXISTS messages (
+        id BLOB PRIMARY KEY,
+        timestamp INTEGER NOT NULL,
+        contentTopic INTEGER NOT NULL, 
+        payload BLOB
+    ) WITHOUT ROWID;
+    """
+
+  ok(MessageStore(
+    env: env.release
+  ))
+
+template prepare(env: Sqlite, q: string, cleanup: untyped): ptr sqlite3_stmt =
+  var s: ptr sqlite3_stmt
+  checkErr sqlite3_prepare_v2(env, q, q.len.cint, addr s, nil):
+    cleanup
+  s
+
+proc bindParam(s: RawStmtPtr, n: int, val: auto): cint =
+  when val is openarray[byte]|seq[byte]:
+    if val.len > 0:
+      sqlite3_bind_blob(s, n.cint, unsafeAddr val[0], val.len.cint, nil)
+    else:
+      sqlite3_bind_blob(s, n.cint, nil, 0.cint, nil)
+  elif val is int32:
+    sqlite3_bind_int(s, n.cint, val)
+  elif val is uint32:
+    sqlite3_bind_int(s, int(n).cint, int(val).cint)
+  elif val is int64:
+    sqlite3_bind_int64(s, n.cint, val)
+  else:
+    {.fatal: "Please add support for the 'kek' type".}
+
+proc put*(db: MessageStore, message: WakuMessage): MessageStoreResult[void] =
+  let s = prepare(db.env, "INSERT INTO messages (id, timestamp, contentTopic, payload) VALUES (?, ?, ?, ?);"): discard
+  checkErr bindParam(s, 1, @(message.id().data))
+  checkErr bindParam(s, 2, toUnix(now().toTime))
+  checkErr bindParam(s, 3, message.contentTopic)
+  checkErr bindParam(s, 4, message.payload)
+
+  let res =
+    if (let v = sqlite3_step(s); v != SQLITE_DONE):
+      err($sqlite3_errstr(v))
+    else:
+      ok()
+
+  # release implict transaction
+  discard sqlite3_reset(s) # same return information as step
+  discard sqlite3_clear_bindings(s) # no errors possible
+
+  res
+
+proc close*(db: MessageStore) =
+  discard sqlite3_close(db.env)
+
+  db[] = MessageStore()[]
+
+proc get*(db: MessageStore, topics: seq[ContentTopic], paging: PagingInfo, onData: DataProc): MessageStoreResult[bool] =
+  var limit = ""
+  if paging.pageSize > 0:
+    limit = "LIMIT " & $paging.pageSize
+  if paging.pageSize > MaxPageSize:
+    limit = "LIMIT " & $MaxPageSize
+
+  let direction =
+    case paging.direction
+    of FORWARD:
+      "timestamp > " & $paging.cursor.receivedTime & " AND id > ?"
+    of BACKWARD:
+      "timestamp < " & $paging.cursor.receivedTime & " AND id < ?"
```

Comment:
The combination of $paging.cursor.receivedTime + paging.cursor.digest must be compared but not each of them individually.  Maybe, its better to find the message corresponding to $paging.cursor.receivedTime and  paging.cursor.digest in the sorted database , and then apply the `limit` depending on the direction

Reply:
btw, if a message corresponding to  $paging.cursor.receivedTime + paging.cursor.digest cannot be found in the database, it is an invalid cursor, an no waku message shall be returned

Reply:
Also, if `pageSize` is not zero AND the cursor has the default value i.e., `paging.cursor.receivedTime=0` and `paging.cursor.digest `is an array of zero bytes, this corresponds to an initial paging request. If `direction` is `FORWARD`, slice from the beginning of the database, and if `BACKWARD` slice from the end of the database.

Reply:
@staheri14 whats the compelling rationale for having implemented it like that in the first place?

Reply:
The value of `pageSize` is used as the indicator of whether the pagination should take place or not. We could alternatively define a new boolean field in the `HistoryQuery` to capture the same thing (true for pagination and false for no pagination), though we can infer the same thing by just looking at the `pageSize` as well.
In fact, if the querying node is not aware of pagination at all and leaves the `pagingInfo` part of the query (hence the `pageSize`) unspecified, everything will work correctly (the queried node can understand that no pagination is required). 
This is how I implemented it in the in-memory version, feel free to follow a different approach :)

Reply:
Also, for the initial pagination request, the querying node does not have a valid `cursor`, that is why the queried node  performs somewhat a default pagination. Regarding the default pagination policy i.e., slicing messages from the beginning or the end of the database, the rationale was that when the requested paging `direction` is `FORWARD`, the user is willing to read messages from the beginning to the end, and vice versa when the direction is `BACKWARD`.

Reply:
I meant why are we doing $paging.cursor.receivedTime + digest?

Reply:
I see! the full discussion is in this issue https://github.com/vacp2p/specs/issues/195, but the overall idea is that, nodes are interested in browsing messages based on the time (usually the most recent ones), that is why messages have timestamp to allow pagination by time. The hash of messages i.e., digests are added to break the tie between two messages that may happen to arrive simultaneously. 
I also think that having hash in the cursor will also allow content-addressing of messages (for the future cases). The querying node can use the hash part of the cursor (obtained  from one node) to resume its history message request by contacting another node.

Also as a side note, the fields of cursor correspond to all the dimensions based on which the pagination is done.

By the way, as the in-memory version is going to be replaced by the database (hence reworking the pagination), we may also modify the pagination method and  use a sender-generated timestamp instead (mentioned in https://github.com/vacp2p/specs/issues/219). What do you think @oskarth

Reply:
Well what this implementation now has is `time` and `cursor` so wouldn't sort by time and then by cursor eg ID yield the result we wanted?

Reply:
Consider this example of two waku messages, 
w1, timetamp=2, digest=1

Reply:
If I understood your implementation correctly, then the following made-up example would clarify my point (w1 and w2 are waku messages)
W1, timestamp=2, hash=10
W2, timestamp=3, hash=9

W1 precedes w2 in the sorted database

Consider a cursor with the following fields: timestamp=2 and hash=10
If you search for messages whose hash AND timestamp are larger than the cursor then you will miss W2 because cursor.hash is not larger than w2.hash

Reply:
Ah I see what you are saying, so instead of it being a WHERE clause we should find all messages greater than timestamp ordered by the hash?

Reply:
I think the hash values should never be treated alone, hashes always make sense in conjunction with timestamp. Maybe it is better to create a temporary column `new_col` from the concatenation of` timestamp` and `hash`, then sort based on `new_col`, and then use the WHERE clause as `WHERE new_col>cursor.timestamp||cursor.digest`

Reply:
Regarding my prior comment, we should make sure that there is a record for the queried cursor (to validate the cursor).

Reply:
i don't think this actually makes sense

Reply:
You don't really have to create another column, the concatenation should happen as part of the sql query that you make

Reply:
I'm saying the way it all works seems unintuitive

Reply:
Sorry I missed your last comment, I would be glad to clarify, let me know which parts are unintuitive?